scripts/visualize/visualize_utils.R

```diff
@@ -0,0 +1,160 @@
+    plot(clipped.sp, ..., xlim = xlim, ylim = ylim)
+  })
+  
+  sp.geoms <- tail(xml2::xml_children(rendered), length(sp))
+  
+  # here either strip the important attributes out and re-add them with a xml_set_attrs call, or lapply the nodeset and add attrs one by one:
+  
+  return(sp.geoms) # removing the <rect/> element...
+}
+
+#' extract the plotting limits from a spatial object, given a sized svg view
+#' 
+#' @param sp a spatial object
+#' @param ... additional arguments passed to the plotting methods of `sp` (e.g., `expandBB`)
+#' @param width the width (in inches) of the svg view
+#' @param height the height (in inches) of the svg view
+#' @param pointsize number of pixels per inch
+#' @param return what limits to return
+#' 
+#' @return a list or numeric vector, depending on what is specified for `return`
+
+get_sp_lims <- function(sp, ..., width = 10, height = 8, pointsize = 12, return = c('xlim','ylim')){
+  bb <- get_sp_bbox(sp)
+  # now default plot
+  # extract usr, return lims from usr
+  .fun <- svglite::svgstring(width = width, height = height, pointsize = pointsize, standalone = F)
+  suppressWarnings(plot(bb, ...)) # warning is for expandBB param, if used
+  usr <- par('usr')
+  dev.off()
+  xlim = usr[c(1,2)]
+  ylim = usr[c(3,4)]
+  
+  list.out <- list(xlim = xlim, ylim = ylim)
+  
+  if (length(return) > 1){
+    return(list.out)
+  } else {
+    return(list.out[[return]])
+  }
+  
+}
+
+#' retain data and features for all polys/points within the plotting range,
+#' while removing those outsite and truncating the boundaries of 
+#' polygons to the bounds of the plotting range. 
+#' 
+#' @param sp a spatial object
+#' @param xlim the x limits of the plot using the same coordinate system as `sp`
+#' @param ylim the y limits of the plot using the same coordinate system as `sp`
+#' @param ... additional args, e.g., `clip.fun` for `Spatial` class method (others unused)
+#' 
+#' @return a clipped sp object
+#' @export
+clip_sp <- function(sp, xlim, ylim, ...) {
+  message('clip_sp is not fully tested for all sp classes')
+  UseMethod(generic = 'clip_sp', object = sp)
+}
+
+
+clip_sp.SpatialPolygonsDataFrame <- function(sp, xlim, ylim, ...){
+  
+  clipped.sp <- NextMethod('clip_sp', sp, ..., clip.fun = rgeos::gIntersection)
+  return(clipped.sp)
+}
+
+clip_sp.SpatialPointsDataFrame <- function(sp, xlim, ylim, ...){
+  
+  clipped.sp <- NextMethod('clip_sp', sp, ..., clip.fun = rgeos::gContains)
+  return(clipped.sp)
+}
+
+clip_sp.Spatial <- function(sp, xlim, ylim, ..., clip.fun = rgeos::gIntersection){
+  
+  clip <- as.sp_box(xlim, ylim, CRS(proj4string(sp)))
+  g.i <- rgeos::gIntersects(sp, clip, byid = T) 
+  
+  # from https://stackoverflow.com/questions/13982773/crop-for-spatialpolygonsdataframe
+  out <- vector(mode="list", length=length(which(g.i)))
+  ii <- 1
+  for (i in seq(along=g.i)) if (g.i[i]) {
+    out[[ii]] <- clip.fun(sp[i,], clip)
+    row.names(out[[ii]]) <- row.names(sp)[i]; ii <- ii+1
+  }
```

Comment:
if you wanted, you could replace lines 144-149 with:
```
out <- lapply(which(g.i), function(i) {
    g.out <- clip.fun(sp[i,], clip)
    row.names(g.out) <- row.names(sp)[i]
    return(g.out)
})
```
which is more complex because uses lapply but less complex because avoids all that i/ii indexing

Reply:
smart.